type_inference/type_retrieval_service.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/python
+#
+# Copyright 2023 Logica Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import cache
+from typing import Dict
+from type_inference.postgresql_type_parser import try_parse_postgresql_type
+import psycopg2
+from os import linesep
+
+from type_inference.type_retrieval_exception import TypeRetrievalException
+
+built_in_types = set()
+
+
+def InitBuiltInTypes(connection_string: str):
+  if built_in_types:
+    return
+
+  with psycopg2.connect(connection_string) as conn:
+    with conn.cursor() as cur:
+        # this SQL query returns all primitive types (defined by PostgreSQL directly)
+        cur.execute('''
+SELECT t.typname as type
+FROM pg_type t
+    LEFT JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
+WHERE n.nspname = 'pg_catalog'
+  AND (
+    t.typrelid = 0 OR (SELECT c.relkind = 'c'
+                       FROM pg_catalog.pg_class c
+                       WHERE c.oid = t.typrelid)
+  )
+  AND NOT EXISTS(
+    SELECT 1
+    FROM pg_catalog.pg_type el
+    WHERE el.oid = t.typelem
+      AND el.typarray = t.oid
+  );''')
+        built_in_types.update((t[0] for t in cur.fetchall()))
+
+
+@cache
```

Comment:
Feels like we may actually need the content of the cache for something. Please do caching manually. It's just a couple of extra lines, but we don't have to think about where is it cached if we need to access.